dig/dig.go

```diff
@@ -211,25 +211,27 @@ func (g *Graph) registerConstructor(c interface{}) error {
 // When a new constructor is being inserted, detect any present cycles
 func (g *Graph) detectCycles(n *funcNode) error {
 	visited := make(map[string]bool)
-	return g.recursiveDetectCycles(n, visited)
+	l := list.New()
+	return g.recursiveDetectCycles(n, visited, l)
 }
 
 // DFS and tracking if same node is visited twice
-func (g *Graph) recursiveDetectCycles(n graphNode, visited map[string]bool) error {
-	if visited[n.id()] {
-		return errCycle
-	}
-
+func (g *Graph) recursiveDetectCycles(n graphNode, visited map[string]bool, l *list.List) error {
 	visited[n.id()] = true
+	l.PushBack(n.id())
```

Comment:
I was pretty sure i needed to pop, but couldn't write a proper case for it on friday evening heh. I will change the test to break (your example) and then fix the array handling